core/controllers/suggestion.py

```diff
@@ -264,17 +264,95 @@ def post(self) -> None:
+                suggestion,
+                suggestion_registry.SuggestionTranslateContent
+            )
+            self._copy_images_from_target_exploration_content_to_translation(
+                suggestion
+            )
+
             files = self.normalized_payload.get('files')
             new_image_filenames = (
                 suggestion.get_new_image_filenames_added_in_suggestion()
             )
             if new_image_filenames and files is not None:
-                _upload_suggestion_images(
-                    files, suggestion, new_image_filenames
+                new_image_files = {
+                    filename: image_blob
+                    for filename, image_blob
+                    in files.items()
+                    if filename in new_image_filenames
+                }
+                self._save_new_images_added_in_translation(
+                    new_image_files, suggestion
                 )
 
         self.render_json(self.values)
 
+    def _save_new_images_added_in_translation(
+        self,
+        new_files: Dict[str, str],
+        suggestion: suggestion_registry.SuggestionTranslateContent
+    ) -> None:
+        """Saves new images introduced in translation suggestion to storage.
+
+        Args:
+            new_files: dict. A mapping from each new image's filename to its
+                corresponding image blob.
+            suggestion: SuggestionTranslateContent. The translation suggestion
+                for which images are being uploaded.
+        """
+        for filename, image in new_files.items():
+            decoded_image = base64.decodebytes(image.encode('utf-8'))
+            file_format = (
+                image_validation_services.validate_image_and_filename(
+                    decoded_image, filename))
+            image_is_compressible = (
+                file_format in feconf.COMPRESSIBLE_IMAGE_FORMATS)
+            fs_services.save_original_and_compressed_versions_of_image(
+                filename, suggestion.image_context, suggestion.target_id,
+                decoded_image, 'image', image_is_compressible)
+
+    def _copy_images_from_target_exploration_content_to_translation(
+        self,
+        suggestion: suggestion_registry.SuggestionTranslateContent
+    ) -> None:
+        """Creates copies of images from the suggestion's target exploration
+        for the translation suggestion to use.
+
+        Args:
+            suggestion: SuggestionTranslateContent. The translation suggestion
+                to copy its target exploration's images to.
+
+        Raises:
+            Exception. An image in the target exploration's content is not a
+                saved asset belonging to the target exploration.
+        """
+        target_image_filenames = (
+            html_cleaner.get_image_filenames_from_html_strings(
+                suggestion.get_target_entity_html_strings()
+            )
+        )
+        try:
+            fs_services.copy_images(
```

Comment:
Just a question about this. Here, target_image_filenames is looking at the source content and copying those images over (let me know if I understood this wrong). But shouldn't we be looking at the new content (the suggestion HTML) to determine which images need to be stored with the suggestion?

Some of those images will be new and some will be existing; we should only copy over the existing ones. I.e. if an image in the source content is replaced then we shouldn't need to store it with the suggestion.

Reply:
> Here, target_image_filenames is looking at the source content and copying those images over (let me know if I understood this wrong).

This is correct.

> But shouldn't we be looking at the new content (the suggestion HTML) to determine which images need to be stored with the suggestion?

There are 2 scenarios that I should outline about the image pipeline:
- Suggestion with image in source content and with uploaded image in new content: We normally copy the image in the source content anyways, as the translation suggestion review modal will need the image to fully display the source content.
- Suggestion with image in source content and with copied image in new content: We should copy the image from the source content, as not only will the translation suggestion review modal need it to fully display the source content, but also to fully display the new content

So in either scenario, we should still be copying the image from the source content. There's not much need to look at the new content to determine which images are existing, as the source content contains only existing images. Also, the suggestion domain model gets only the new images from the new content already.

> I.e. if an image in the source content is replaced then we shouldn't need to store it with the suggestion.

I'm not sure what you mean by "replaced". Are you referring to the exploration creator changing the image in the target exploration, or are you referring to the copied image in the new content "replacing" the image in the source content?

Reply:
> Suggestion with image in source content and with uploaded image in new content: We normally copy the image in the source content anyways, as the translation suggestion review modal will need the image to fully display the source content.

Hm, this part seems a bit odd to me. Shouldn't the review modal get the source images from the source content URL and the translation images from the suggestion content URL? (If the source content changes between the time of the suggestion and the time of review, we would show the latest source content, right -- so technically the source content isn't a part of the suggestion?)

> I'm not sure what you mean by "replaced". Are you referring to the exploration creator changing the image in the target exploration, or are you referring to the copied image in the new content "replacing" the image in the source content?

The former.

Reply:
> Hm, this part seems a bit odd to me. Shouldn't the review modal get the source images from the source content URL and the translation images from the suggestion content URL? (If the source content changes between the time of the suggestion and the time of review, we would show the latest source content, right -- so technically the source content isn't a part of the suggestion?)

@seanlip That's not what's happening on the `develop` branch.

When the translation submitter submits a translation whose source content contains an image, the image is copied from GCS `/exploration/` to `/exploration_suggestions/`.

Afterwards, when a relevant translation reviewer reviews the submitted translation suggestion, the image within the source content is fetched from GCS `/exploration_suggestions`.

The results in my [debugging doc "Hypothesis testing" section](https://docs.google.com/document/d/1sL74bmX1A7YLGrbLqY0yAuByZk3nCeIJl65V8sMmINI/edit#heading=h.fu7cskbxenio), specifically on the table cell at the 4th row from the top and 2nd column from the left, proves this, at least for the state of `develop` that was 1-2 months ago.

Reply:
Oh, I see. But we know the flow in the develop branch is kinda broken. Conceptually, shouldn't the source text be taking images from the source folder? (Otherwise, for example, if new images are added to the source between submission and review, they won't show up here.)

Reply:
@seanlip That's a good point. I'm wondering if I can address this in a separate issue (preferably the next high-priority issue I can work on). 

The goal of the linked issue (and by extension, this PR) is to make images created by the copy tool to load properly. This seems different from the goal of keeping image-related updates to source content in-sync with corresponding submitted suggestions.

Also, I'm not sure if I can find all relevant parts of the codebase, make the appropriate changes to them, and then address all reviewer comments that spawn from these changes, all within 3 days (since I'm trying to get this merged by April 30th).

Reply:
@StephenYu2018 Yup that's fine (to address in a separate issue) ... could you please file a follow-up issue and link it here? Then I'll approve this and we can merge it.

Thanks!

Reply:
@seanlip I've filed this concern at the follow-up issue #20232.